### ISSUE TITLE
Rewrite cmake script

### ISSUE THREAD

**Issue by sytelus** [rpclib/rpclib#94]
CMake is building 32-bit build even when executed from "Visual Studio 2015 x64 native Tools" command prompt. I'd to do following to get 64-bit build. It might be nice to include this in documentation.

1. Go to "Visual Studio 2015 x64 native Tools" command prompt. Make sure you have VS2015 Update 3.
2. `cmake -G"Visual Studio 14 2015 Win64"  ..`
3. `cmake --build .`

This will generate Debug build. To get the release build do `cmake --build . --config Release`.

**Comment by sztomi** [comment]
AFAIK this is default CMake bahavior, but the doc addition is a nice idea. I'll add it for the release, thanks!

**Comment by sytelus** [comment]
I think CMake is supposed to detect that its x64 environment and automatically do x64 build. Generally if I invoke CMake from VS2015 x64 Native Command Prompt then it builds x64 by default. But rpclib didn't do that...

**Comment by sztomi** [comment]
OK, thanks for pointing that out. I'll have to do some investigation then. There *is* some machinery in cmake about detecting the current platform, but maybe this fails on windows.

**Comment by zauberparacelsus** [comment]
I just happened upon this bug report because I was trying to determine how to tell cmake to build 64bit on Windows.  Since I'm having trouble doing that, this report has me interested.

@sztomi I can confirm that cmake is correctly detecting my system as 64bit and compiling binaries as such.  So yeah, it seems there is a problem on Windows.

**Comment by sztomi** [comment]
So I realized that I'm dragging along a lot of bad decisions stemming from my early experimentation with cmake. Since the library is still relatively small, I'm going to rewrite the cmake script and fix this problem as well.

**Comment by concatime** [comment]
Well actually, the result needs to be like this: `cmake -G "Visual Studio 15 2017 Win64" -T host=x64` (or whatever version) if I'm not wrong.
https://cmake.org/cmake/help/v3.8/generator/Visual%20Studio%2015%202017.html#toolset-selection
https://gitlab.kitware.com/cmake/cmake/issues/15622#note_180677

**Comment by AravK** [comment]
About the VERSION variables: `project()` takes `VERSION <major>.<minor.<patch>.<tweak>` off which all are optional. Variables are defined as `PROJECT_VERSION_MAJOR / PROJECT_VERSION_MINOR` ... or `<PROJECT_NAME>_VERSION_MAJOR` ... Read the CMake 3.8 documentation on [`project`](https://cmake.org/cmake/help/v3.8/command/project.html)

**Comment by ahoarau** [comment]
I also need to explicitly add the generator for building things on windows. There are so many environments (native, cygwin32,64, minsys2 32,64 ...). I don't think it is special to this particular lib not being able to detect x64 hosts. 
Adding this to the wiki is certainly useful

**Comment by sztomi** [comment]
For anyone interested, I started working on this (finally!).

**Comment by mharinder** [comment]
Have we considered adding this to [vcpkg](https://github.com/Microsoft/vcpkg) ? I think that'll require some cmake changes. I have

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index bd605eb..c3391af 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -229,7 +229,7 @@ set(DEP_SOURCES
     dependencies/src/format.cc
     dependencies/src/posix.cc)
 
-add_library(${OUTPUT_LIBRARY_NAME}
+add_library(${OUTPUT_LIBRARY_NAME} SHARED
     lib/rpc/dispatcher.cc
     lib/rpc/server.cc
     lib/rpc/client.cc
@@ -285,7 +285,10 @@ if (RPCLIB_MSVC_STATIC_RUNTIME)
 
 endif()
 
-install(TARGETS ${OUTPUT_LIBRARY_NAME} DESTINATION lib)
+install(TARGETS ${OUTPUT_LIBRARY_NAME}
+	RUNTIME DESTINATION bin
+	ARCHIVE DESTINATION lib
+	LIBRARY DESTINATION lib )
 install(DIRECTORY include/
     DESTINATION include
     FILES_MATCHING
@@ -293,7 +296,7 @@ install(DIRECTORY include/
     PATTERN "*.hpp"
     PATTERN "*.inl"
     PATTERN "*.in" EXCLUDE)
-install(FILES ${PROJECT_BINARY_DIR}/version.h DESTINATION include/rpc)
+install(FILES ${PROJECT_SOURCE_DIR}/include/rpc/version.h DESTINATION include/rpc)
 
 if(RPCLIB_GENERATE_COMPDB)
     set(CMAKE_EXPORT_COMPILE_COMMANDS "ON") # for YCM
```

to make it look more like other vcpkg packages. But I still have difficulty dealing with "dll export" stuff. A quick fix would be passing `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=TRUE`. But that exports too many symbols.

**Comment by sztomi** [comment]
Well I'm not against it, but I don't plan to support vcpkg explicitly. An official conan package is coming, on the other hand.

**Comment by sztomi** [comment]
As far as I can tell, all reported problems are fixed and it's in a much better shape now than it used to be. Release soon. Feel free to test it on the dev branch (1746b0b)

**Comment by sztomi** [comment]
This has been released in [v2.2.0](https://github.com/rpclib/rpclib/releases/tag/v2.2.0). Please update.

**Comment by ras0219-msft** [comment]
@mharinder rpclib v2.2.0 is now available in vcpkg.

@sztomi thanks for rewriting your CMake, it's a lot better!
